genericvalidator/acceptable_values_validator.go

```diff
@@ -0,0 +1,82 @@
+package genericvalidator
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/helpers/validatordiag"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/tfsdk"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+// acceptableValuesAttributeValidator is the underlying struct implementing OneOf and NoneOf.
+type acceptableValuesAttributeValidator struct {
+	acceptableValues []attr.Value
+	shouldMatch      bool
+}
+
+var _ tfsdk.AttributeValidator = (*acceptableValuesAttributeValidator)(nil)
+
+func (av *acceptableValuesAttributeValidator) Description(ctx context.Context) string {
+	return av.MarkdownDescription(ctx)
+}
+
+func (av *acceptableValuesAttributeValidator) MarkdownDescription(_ context.Context) string {
+	if av.shouldMatch {
+		return fmt.Sprintf("Value(s) must match one of: %q", av.acceptableValues)
+	} else {
+		return fmt.Sprintf("Value(s) must match none of: %q", av.acceptableValues)
+	}
+
+}
+
+func (av *acceptableValuesAttributeValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, res *tfsdk.ValidateAttributeResponse) {
+	if req.AttributeConfig.IsNull() || req.AttributeConfig.IsUnknown() {
+		return
+	}
+
+	// Gather the `values` to validate:
+	// singleton slice for primitives,
+	// a collection for the other possible `types.*`.
```

Comment:
This is nice and DRY and removes the need to have `OneOf` and `NoneOf` on each of the type validators (e.g., `mapvalidator`, `float64validator`). Perhaps it's not a concern, but having a `genericvalidator` means that we're not checking the type, for instance, `OneOf` could pass validation if a `Set` was supplied, when a `List` should have been used. In the specific validators (e.g., `float64.AtLeast`) there is an initial check that the value being validated is of the expected type. Just thought I'd mention this.